### PR TITLE
Rewrite core RX services initial

### PR DESCRIPTION
**Pull request by bhavesh-desai-scratch**
[WIP]

Part of an ongoing work to bring Reactive trader up to date. Focus on is on the core RXJS layer.
- Rewrite core RXJS services to make them easier to understand
- Increase use of types throughout the application
- Stop using inner observables
- Use refcount instead of publish where possible
- Improve logging
- Stop subscribing to status multiple times

### CODE REVIEW

File: src/client/src/services/connectionStatusService.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          } else {
+            return {
+              status: ConnectionStatus.disconnected,
+              url: 'Disconnected',
+              transportType: ConnectionType.Unknown
+            }
+          }
+        }
+      }),
+      distinctUntilChanged(),
```
Comment (fdore) [inline src/client/src/services/connectionStatusService.ts]: I don't think distinctUntilChanged will have any effect, given the map returns a new reference every time we get a connection update.
You might need to pass the predicate `(a, b) => a.status === b.status && a.url === b.url && a.transportType === b.transportType`

File: src/client/src/system/service/connection.ts [hunk 2]
```diff
@@ -141,62 +76,75 @@ export class Connection {
...
-          }
-        } catch (err) {
-          log.error(`Error thrown unsubscribing from topic ${topic}`, err)
-        }
-      }
-    })
+
+            return () => {
+              log.info(`Tearing down topic ${topic}`)
+              //obs.complete()
```
Comment (fdore) [inline src/client/src/system/service/connection.ts]: can remove comments

File: src/client/src/system/service/connection.ts [hunk 3]
```diff
@@ -208,46 +156,41 @@ export class Connection {
...
+                obs.unsubscribe()
+              },
+              error => {
+                obs.error(error)
+              }
+            )
+
+            return () => {
+              obs.complete()
+              // obs.unsubscribe()
```
Comment (fdore) [inline src/client/src/system/service/connection.ts]: can remove

File: src/client/src/system/service/connection.ts [hunk 4]
```diff
@@ -208,46 +156,41 @@ export class Connection {
...
+      mergeMap(
+        ({ session }) =>
+          new Observable<TResult>(obs => {
+            log.debug(`Doing a RPC to [${remoteProcedure}]]`)
+
+            const dto: SubscriptionRequest<TPayload> = [
+              {
+                payload,
+                replyTo: responseTopic,
+                Username: this.userName
```
Comment (fdore) [inline src/client/src/system/service/connection.ts]: any reason for the different casing?

File: src/client/src/system/service/serviceClient.ts [hunk 5]
```diff
@@ -172,196 +41,101 @@ export default class ServiceClient extends DisposableBase {
...
-      // request-response and stream operations as is currently the case.
-      // tslint:disable-next-line:no-bitwise
-      const topicName = `topic_${this.serviceType}_${
+    return this.getServiceWithMinLoad$(service).pipe(
+      take(1),
+      switchMap(serviceInstanceStatus => {
+        // The backend has a different contract for streams (i.e. request-> n responses) as it does with request-response (request->single response) thus
+        // the different method here to support this.
+        // It works like this: client creates a temp topic, we perform a RPC to then tell the backend to push to this topic.
+        // TBH this is a bit odd as the server needs to handle fanout and we don't have any really control over the attributes of the topic, however it's sufficient for our needs now.
```
Comment (fdore) [inline src/client/src/system/service/serviceClient.ts]: don't have any really control => don't have any _real_ control

File: src/client/src/system/service/serviceClient.ts [hunk 6]
```diff
@@ -172,196 +41,101 @@ export default class ServiceClient extends DisposableBase {
...
-    serviceId: serviceStatus.Instance,
-    timestamp: serviceStatus.TimeStamp,
-    serviceLoad: serviceStatus.Load,
-    isConnected: true
-  }
-}
+                .pipe(take(1))
+                .subscribe(ack => {
+                  this.log.info(`request acknowledged for ${remoteProcedure}`)
+                  req.unsubscribe()
```
Comment (fdore) [inline src/client/src/system/service/serviceClient.ts]: you dont need to unsubscribe as you are taking 1

File: src/client/package.json [hunk 7]
```diff
@@ -96,22 +99,25 @@
...
-    "lint-staged": "^7.0.5",
+    "lint-staged": "^7.1.0",
     "node-sass-chokidar": "^1.2.2",
     "npm-run-all": "^4.1.2",
     "openfin-cli": "1.1.5",
     "prettier": "^1.12.1",
     "react-scripts-ts": "2.15.1",
     "react-test-renderer": "^16.3.2",
     "redux-devtools-extension": "2.13.2",
+    "rxjs-compat": "^6.1.0",
```
[the remark this answers is not in the document]
Comment (bhavesh-desai-scratch) [reply src/client/package.json]: This is actually a dev dep as it is only used for one of the build tools. 'lint-staged'. The actual application is on RxJS 6

File: src/client/src/services/connectionStatusService.ts (hunk 1 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/services/connectionStatusService.ts]: Good spot. Was left over from old code

File: src/client/src/services/executionService.ts [hunk 8]
```diff
@@ -32,92 +28,83 @@ const log = logger.create('ExecutionService')
...
-  openFin: any
-) {
-  const serviceClient = new ServiceClient(
-    ServiceConst.ExecutionServiceKey,
-    connection
-  )
+export default class ExecutionService {
+  constructor(
+    private readonly serviceClient: ServiceClient,
+    private readonly openFin: any
```
[the remark this answers is not in the document]
Comment (bhavesh-desai-scratch) [reply src/client/src/services/executionService.ts]: done

File: src/client/src/services/referenceDataService.ts [hunk 9]
```diff
@@ -1,132 +1,51 @@
...
   UpdateType
 } from '../types'
 import { referenceDataMapper } from './mappers'
 import { RawCurrencyPairUpdates } from './mappers/referenceDataMapper'
 
-type CurrencyPairCache = (update: CurrencyPairUpdates) => void
-
-const log = logger.create('ReferenceDataService')
+export default class ReferenceDataService {
+  private readonly referenceDataStreamConnectable: Observable<
```
[the remark this answers is not in the document]
Comment (bhavesh-desai-scratch) [reply src/client/src/services/referenceDataService.ts]: done

File: src/client/src/system/service/connection.ts (hunk 2 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/system/service/connection.ts]: done

File: src/client/src/system/service/connection.ts (hunk 3 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/system/service/connection.ts]: done

File: src/client/src/system/service/connection.ts (hunk 4 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/system/service/connection.ts]: The server takes the uppercase version, So We convert it at the last possible moment

File: src/client/src/system/service/serviceClient.ts (hunk 5 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/system/service/serviceClient.ts]: Done

File: src/client/src/system/service/serviceClient.ts (hunk 6 above)
Comment (bhavesh-desai-scratch) [reply src/client/src/system/service/serviceClient.ts]: Ha. Good spot again

**Pull request merged**